## 异常

## Exception和Error区别

* **`Exception`** :程序本身可以处理的异常，可以通过 `catch` 来进行捕获。
* **`Error`** ：`Error` 属于程序无法处理的错误 ，不建议通过 `catch`捕获 。例如 Java 虚拟机运行错误（`Virtual MachineError`）、虚拟机内存不够错误(`OutOfMemoryError`)、类定义错误（`NoClassDefFoundError`）等 。这些异常发生时，Java 虚拟机（JVM）一般会选择线程终止。

## Checked Exception 和 Unchecked Exception 区别

* **`Checked Exception`** 即 受检查异常 ，Java 代码在编译过程中，如果受检查异常没有被 `catch`或者 `throws` 关键字处理的话，就没办法通过编译。
* **`Unchecked Exception`** 即 不受检查异常 ，Java 代码在编译过程中 ，我们即使不处理不受检查异常也可以正常通过编译。

## throw和throws

* throws 关键字用于声明异常，它的作用和 try-catch 相似；而 throw 关键字用于显式的抛出异常。
* throws 关键字后面跟的是异常的名字；而 throw 关键字后面跟的是异常的对象。

## try-catch-finally

* finally 块前面必须有 try 块，不要把 finally 块单独拉出来使用。编译器也不允许这样做。
* finally 块不是必选项，有 try 块的时候不一定要有 finally 块。
* 如果 finally 块中的代码可能会发生异常，也应该使用 try-catch 进行包裹。
* 即便是 try 块中执行了 return、break、continue 这些跳转语句，finally 块也会被执行。
* 不执行finally的情况：
  * 遇到了死循环。
  * 执行了 `System. exit()` 这行代码。

## try-with-resources

* 取消finally块，把需要释放的资源写在 try 后的 `()` 中
* 如果想释放自定义资源的话，让它实现 AutoCloseable 接口，并提供 `close()` 方法

## 异常处理实践经验

* 尽量不要捕获RuntimeException
* 尽量使用 try-with-resource 来关闭资源，禁止在 try 块中直接关闭资源。
* 不要捕获 Throwable
* 不要省略异常信息的记录
* 不要记录了异常又抛出异常
* 不要在finally块中使用return
* 抛出具体定义的检查性异常而不是 Exception
* 捕获具体的子类而不是捕获 Exception 类
* 自定义异常时不要丢失堆栈跟踪
* finally 块中不要抛出任何异常
* 不要在生产环境中使用 ` printStackTrace()`
* 对于不打算处理的异常，直接使用 try-finally，不用 catch
* 记住早 throw 晚 catch 原则
* 只抛出和方法相关的异常
* 切勿在代码中使用异常来进行流程控制
* 尽早验证用户输入以在请求处理的早期捕获异常
* 一个异常只能包含在一个日志中
* 将所有相关信息尽可能地传递给异常
* 终止掉被中断线程
* 对于重复的 try-catch，使用模板方法
